### PR TITLE
Modernize generated code for Hydrators

### PR DESCRIPTION
**Pull request by GromNaN**
|      Q       |   A
|------------- | -----------
| Type         | improvement
| BC Break     | no
| Fixed issues | -

- Add `use function` statement for compiler-optimized functions
- Using null coalescence and short array syntax
- Adding property type with CPP.

### CODE REVIEW

File: lib/Doctrine/ODM/MongoDB/Hydrator/HydratorFactory.php [hunk 1]
```diff
@@ -337,7 +337,7 @@ private function generateHydratorClass(ClassMetadata $class, string $hydratorCla
                 $this->dm->getUnitOfWork()->setParentAssociation($return, $this->class->fieldMappings['%2$s'], $document, '%1$s');
 
                 $embeddedData = $this->dm->getHydratorFactory()->hydrate($return, $embeddedDocument, $hints);
-                $embeddedId = $embeddedMetadata->identifier && isset($embeddedData[$embeddedMetadata->identifier]) ? $embeddedData[$embeddedMetadata->identifier] : null;
+                $embeddedId = $embeddedMetadata->identifier && $embeddedData[$embeddedMetadata->identifier] ?? null;
```
Comment (GromNaN) [inline lib/Doctrine/ODM/MongoDB/Hydrator/HydratorFactory.php]: Using null coalescence.

File: lib/Doctrine/ODM/MongoDB/Hydrator/HydratorFactory.php [hunk 2]
```diff
@@ -370,23 +370,20 @@ private function generateHydratorClass(ClassMetadata $class, string $hydratorCla
 use Doctrine\ODM\MongoDB\Query\Query;
 use Doctrine\ODM\MongoDB\Mapping\ClassMetadata;
 
+use function array_key_exists;
+use function gettype;
+use function is_array;
```
Comment (GromNaN) [inline lib/Doctrine/ODM/MongoDB/Hydrator/HydratorFactory.php:375]: This functions are compiler-optimized.

File: lib/Doctrine/ODM/MongoDB/Hydrator/HydratorFactory.php [hunk 3]
```diff
@@ -370,23 +370,20 @@ private function generateHydratorClass(ClassMetadata $class, string $hydratorCla
...
 {
-    private \$dm;
-    private \$class;
-
-    public function __construct(DocumentManager \$dm, ClassMetadata \$class)
-    {
-        \$this->dm = \$dm;
-        \$this->class = \$class;
-    }
+    public function __construct(private DocumentManager \$dm, private ClassMetadata \$class) {}
```
Comment (GromNaN) [inline lib/Doctrine/ODM/MongoDB/Hydrator/HydratorFactory.php:382]: Adding the property type.

File: lib/Doctrine/ODM/MongoDB/Hydrator/HydratorFactory.php [hunk 4]
```diff
@@ -370,23 +370,20 @@ private function generateHydratorClass(ClassMetadata $class, string $hydratorCla
...
 
-    public function hydrate(object \$document, array \$data, array \$hints = array()): array
+    public function hydrate(object \$document, array \$data, array \$hints = []): array
     {
-        \$hydratedData = array();
+        \$hydratedData = [];
```
Comment (GromNaN) [inline lib/Doctrine/ODM/MongoDB/Hydrator/HydratorFactory.php:386]: Short-array syntax.

File: lib/Doctrine/ODM/MongoDB/Hydrator/HydratorFactory.php [hunk 5]
```diff
@@ -167,8 +167,8 @@ private function generateHydratorClass(ClassMetadata $class, string $hydratorCla
                     $code .= sprintf(
                         <<<EOF
 
-        /** @AlsoLoad("$name") */
-        if (!array_key_exists('%1\$s', \$data) && array_key_exists('$name', \$data)) {
+        // AlsoLoad("$name")
```
Comment (GromNaN) [inline lib/Doctrine/ODM/MongoDB/Hydrator/HydratorFactory.php:170]: Reference to doctrine annotations is outdated.

**Review by malarzm — APPROVED** [review]

**Pull request merged**